src/hash.c

```diff
@@ -1111,6 +1111,52 @@ mrb_hash_foreach(mrb_state *mrb, struct RHash *h, mrb_hash_foreach_func *func, v
   });
 }
 
+MRB_API mrb_hash_iterator
+mrb_hash_iterator_new(struct RHash *h)
+{
+  mrb_hash_iterator it;
+
+  if (h_ar_p(h)) {
+    it.entry = ar_ea(h);
+    it.rem = ar_size(h);
+  } else {
+    it.entry = ht_ea(h);
+    it.rem = ht_size(h);
+  }
+
+  if (0 < it.rem) {
+    while (entry_deleted_p(it.entry)) {
+      ++it.entry;
+    }
+  }
+
+  return it;
+}
+
+MRB_API mrb_bool
+mrb_hash_iterator_move_next(mrb_hash_iterator *it)
+{
+  if (0 < it->rem) {
+    while (entry_deleted_p(++it->entry));
```

Comment:
If `it->rem` is 1, then `++it->entry` will cause a buffer overrun.
This may be the reason why all VisualC++ 32-bit versions by appveyor are failing.

Reply:
I'll look into that. I may have to change the API slightly to make it safer. Will report back soon.

Reply:
I'm sure there are other ways, but as a point of reference, I made the following changes and valgrind no longer reports problems here.

```patch
diff --git a/src/hash.c b/src/hash.c
index eb049d0e7..12b5e0215 100644
--- a/src/hash.c
+++ b/src/hash.c
@@ -1136,10 +1136,9 @@ mrb_hash_iterator_new(struct RHash *h)
 MRB_API mrb_bool
 mrb_hash_iterator_move_next(mrb_hash_iterator *it)
 {
-  if (0 < it->rem) {
+  if (0 < it->rem && 0 < --it->rem) {
     while (entry_deleted_p(++it->entry));
-    --it->rem;
-    return it->rem > 0;
+    return TRUE;
   }

   return FALSE;
```